lib/collection/src/shards/replica_set/update.rs

```diff
@@ -130,80 +131,80 @@ impl ShardReplicaSet {
+                "The replica set for shard {} on peer {} has no active replica",
+                self.shard_id, this_peer_id
+            )));
+        }
 
-                    let local_update = async move {
-                        local
-                            .get()
-                            .update(operation, local_wait)
-                            .await
-                            .map(|ok| (this_peer_id, ok))
-                            .map_err(|err| (this_peer_id, err))
-                    };
+        let current_clock_tick = clock.tick_once();
+        let clock_tag = ClockTag::new(this_peer_id, clock.id() as _, current_clock_tick);
+        let operation = OperationWithClockTag::new(operation, Some(clock_tag));
 
-                    update_futures.push(local_update.left_future());
-                }
-            }
+        let mut update_futures = Vec::with_capacity(active_remote_shards.len() + 1);
+
+        if let Some(local) = local.deref() {
+            if self.peer_is_active_or_pending(&this_peer_id) {
+                let local_wait = if self.peer_state(&this_peer_id) == Some(ReplicaState::Listener) {
+                    false
+                } else {
+                    wait
+                };
 
-            for remote in active_remote_shards {
                 let operation = operation.clone();
 
-                let remote_update = async move {
-                    remote
-                        .update(operation, wait)
+                let local_update = async move {
+                    local
+                        .get()
+                        .update(operation, local_wait)
                         .await
-                        .map(|ok| (remote.peer_id, ok))
-                        .map_err(|err| (remote.peer_id, err))
+                        .map(|ok| (this_peer_id, ok))
+                        .map_err(|err| (this_peer_id, err))
                 };
 
-                update_futures.push(remote_update.right_future());
+                update_futures.push(local_update.left_future());
             }
+        }
 
-            match self.shared_storage_config.update_concurrency {
-                Some(concurrency) => {
-                    futures::stream::iter(update_futures)
-                        .buffer_unordered(concurrency.get())
-                        .collect()
-                        .await
-                }
+        for remote in active_remote_shards {
+            let operation = operation.clone();
+
+            let remote_update = async move {
+                remote
+                    .update(operation, wait)
+                    .await
+                    .map(|ok| (remote.peer_id, ok))
+                    .map_err(|err| (remote.peer_id, err))
+            };
+
+            update_futures.push(remote_update.right_future());
+        }
 
-                None => FuturesUnordered::from_iter(update_futures).collect().await,
```

Comment:
It feels like changes in this part of code are not strictly required for the PR.
If it is a refactoring, could we please keep it in separate PRs next time, otherwise it is quite hard to review and make sure new changes + refactoring don't break anything

Reply:
This is _kinda_ needed, cause everything was wrapped in a `{}` block before, but now we need to access `clock` (to update the clock during response handling), and so it has to be out of the `{}` block, so I had to restructure it.

Reply:
And if we just put `let mut clock = self.clock_set.lock().await.get_clock();` outside the `{}` block, then it would be locked before other locks and which may generate too many `clock_id`s. :/

Reply:
As a side effect of this change, we're now not dropping the `local` and `remotes` in time. We should probably drop both explicitly after `let all_res = ...` to prevent holding on to them longer than we need.

Reply:
ok, then we need to add `drop` to simulate prev implementation (commited)

Reply:
> As a side effect of this change, we're now not dropping the local and remotes in time...

Right. That's why I put them into a block in the first place. 🤦‍♀️

Reply:
I think this is resolved now, so, marking it as such. :white_check_mark: